### PR TITLE
Layout elevation profile item groundwork

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR implements the groundwork for a layout "elevation profile" item.

Currently this is a stepping stone toward an elevation profile item only -- it exposes an elevation profile item in layouts but the profiles will only render and expose settings relating to the plot background (ie the chart area and axis).

I'm opening this PR prior to implementing the actual profile rendering in the item because this step is all the "boring" groundwork, which ends up with many lines of code changed but nothing of any complexity or needing in-depth review. (I.e, it will allow me to open PRs for the more complex parts which aren't bogged down by this easy/lengthy stuff)!

While we don't yet get elevation profiles rendered, there's a lot to get excited about here anyway!

Here's all the wonderful settings users can use to control the appearance of their plots (this is all fully working in this PR):

image

image

We have:

- control over axis grid intervals (including data-defined intervals), line symbols for the major and minor grid lines, custom number formats for the grid labels, and text format control for grid labels (using text renderer, so with support for buffers/shadows/html formatting/data defined properties)
- control over the chart background area appearance, using fill symbols
- control over the visible chart ranges, ie the elevation and distance ranges, with data-defined control over these for atlas prints
- control over chart area margins (including data defined control)

All these settings result in a super-powerful, flexible chart item which users can control to get exactly the chart appearance they desire!

Additionally, because the charts are all rendered using native QGIS drawing, the charts are exported as vector objects when exporting layouts, making it easy for users to apply post-process edits after exports.

### CODE REVIEW

**Comment by nyalldawson** [comment]
Some chart junk for your viewing pleasure:

image

**Comment by nicogodet** [comment]
Woooo sweet!

Does spinbox reset value should be set to initial value ? (like it is done when exporting a profile to PNG)

image
image

Resetting all spinbox create an empty plot
image

Is there a way to add a reset "value" for border and background ?
Background is drawn bellow axis and profile and border is drawn above axis and profile, correct ?
Things can become a bit confusing when outline is define in background and fill in border.
Maybe restrict things like it's done for item background and border ?
image

Otherwise, I would love to see all this settings added to Export to PNG and Export to PDF in profile widget !

Nice work !

**Comment by nyalldawson** [comment]
@nicogodet 
> Does spinbox reset value should be set to initial value ? (like it is done when exporting a profile to PNG)

Not yet, but it will when we actually have content in the plot

> Is there a way to add a reset "value" for border and background ?

Yep, will do

> Background is drawn bellow axis and profile and border is drawn above axis and profile, correct ?

Correct

> Things can become a bit confusing when outline is define in background and fill in border.
> Maybe restrict things like it's done for item background and border ?

The issue is that item background/border are rather outdated settings, and only support very simple styling. I'd like to upgrade these at some stage too to support full QGIS symbology, so that eg you could have a gradient background on any item.

I agree that users can get in confusing situations if they customise these symbols to a crazy level, but I'd still rather be more flexible and allow full symbology options.

I'll add the "reset" options to the symbol selectors and see if that helps out with this issue...

> Otherwise, I would love to see all this settings added to Export to PNG and Export to PDF in profile widget !

I'm not planning on doing that -- I think it's better to keep the profile chart export dialogs simpler, and require users to setup a print layout if they want to fine-tune the colors/fonts/appearance of the charts further.

File: src/core/layout/qgslayoutitem.h [hunk 1]
```diff
@@ -182,6 +184,14 @@ class CORE_EXPORT QgsLayoutItem : public QgsLayoutObject, public QGraphicsRectIt
         sipType = sipType_QgsLayoutFrame;
         *sipCppRet = static_cast<QgsLayoutFrame *>( sipCpp );
         break;
+      case QGraphicsItem::UserType + 117:
+        sipType = sipType_QgsLayoutItemMarker;
+        *sipCppRet = static_cast<QgsLayoutItemMarker *>( sipCpp );
```
Comment (nirvn) [inline src/core/layout/qgslayoutitem.h:189]: good catch :)

File: src/core/layout/qgslayoutitemelevationprofile.cpp [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+  : QgsLayoutItem( layout )
+  , mPlot( std::make_unique< QgsLayoutItemElevationProfilePlot >() )
+{
+
+  //get default layout font from settings
+  const QgsSettings settings;
+  const QString defaultFontString = settings.value( QStringLiteral( "LayoutDesigner/defaultFont" ), QVariant(), QgsSettings::Gui ).toString();
+  if ( !defaultFontString.isEmpty() )
+  {
+//    QFont f = mFormat.font();
```
Comment (nirvn) [inline src/core/layout/qgslayoutitemelevationprofile.cpp:53]: This whole commented out block, should it be taken out?

File: src/core/layout/qgslayoutitemelevationprofile.cpp [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgslayoutitemelevationprofile.h"
+#include "qgslayoutitemregistry.h"
+#include "qgsplot.h"
+#include "qgsmessagelog.h"
+
+///@cond PRIVATE
+class QgsLayoutItemElevationProfilePlot : public Qgs2DPlot
```
Comment (nirvn) [inline src/core/layout/qgslayoutitemelevationprofile.cpp:24]: @nyalldawson , I'll be very interested in how you develop this specific class :) I think this is one place where we could share code between the qgsquick stuff and layout.

File: src/core/layout/qgslayoutobject.cpp [hunk 4]
```diff
@@ -109,6 +113,16 @@ void QgsLayoutObject::initPropertyDefinitions()
     { QgsLayoutObject::ScalebarHeight, QgsPropertyDefinition( "dataDefinedScalebarHeight", QObject::tr( "Scalebar height in mm" ), QgsPropertyDefinition::DoublePositive ) },
     { QgsLayoutObject::ScalebarSubdivisionHeight, QgsPropertyDefinition( "dataDefinedScalebarSubdivisionHeight", QObject::tr( "Subdivision height in mm" ), QgsPropertyDefinition::DoublePositive ) },
     { QgsLayoutObject::ScalebarRightSegmentSubdivisions, QgsPropertyDefinition( "dataDefinedScalebarRightSegmentSubdivisions", QObject::tr( "Number of subdivisions in segments to the right of 0" ), QgsPropertyDefinition::IntegerPositive ) },
+    { QgsLayoutObject::ElevationProfileDistanceMajorInterval, QgsPropertyDefinition( "dataDefinedElevationProfileDistanceMajorInterval", QObject::tr( "Major grid line interval" ), QgsPropertyDefinition::DoublePositive ) },
```
Comment (nirvn) [inline src/core/layout/qgslayoutobject.cpp:116]: Is the interval in a fixed unit type? Should the description include a mention of which axis is being handled here? I.e. `Distance axis major grid line interval` ?

File: src/gui/qgssymbolbutton.h [hunk 5]
```diff
@@ -162,6 +162,42 @@ class GUI_EXPORT QgsSymbolButton : public QToolButton
...
+
+    /**
+     * Returns the default symbol for the button, which is shown in the button's drop-down menu for the
+     * "default symbol" option.
+     * \returns default symbol for the button. Returns NULLPTR if the default symbol
+     * option is disabled.
+     * \see setDefaultSymbol()
+     * \since QGIS 3.30
+     */
+    const QgsSymbol *defaultSymbol() const;
```
Comment (nirvn) [inline src/gui/qgssymbolbutton.h:183]: Nice, loving this additional functionality.

File: src/ui/layout/qgslayoutelevationprofilewidgetbase.ui [hunk 6]
```diff
@@ -0,0 +1,703 @@
...
+   <header>qgspropertyoverridebutton.h</header>
+  </customwidget>
+ </customwidgets>
+ <tabstops>
+  <tabstop>scrollArea</tabstop>
+ </tabstops>
+ <resources/>
+ <connections/>
+ <buttongroups>
+  <buttongroup name="buttonGroup_2"/>
```
Comment (nirvn) [inline src/ui/layout/qgslayoutelevationprofilewidgetbase.ui:700]: Irrelevant QtCreator crumbs?

File: src/core/layout/qgslayoutitemelevationprofile.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/layout/qgslayoutitemelevationprofile.cpp:53]: yep, thanks!

File: src/ui/layout/qgslayoutelevationprofilewidgetbase.ui (hunk 6 above)
Comment (nyalldawson) [reply src/ui/layout/qgslayoutelevationprofilewidgetbase.ui]: fixed

**Comment by nyalldawson** [comment]
Thanks @nirvn  -- all resolved!

**Review by nirvn — APPROVED** [review]

**Comment by DelazJ** [comment]
@nyalldawson is this the PR that belongs to the changelog or the documentation repository?

**Pull request merged**